src/stores/SpaceStore.tsx

```diff
@@ -517,37 +519,32 @@ export class SpaceStoreClass extends AsyncStoreWithClient<IState> {
 
                 // Don't auto-switch rooms when reacting to a context-switch
                 // as this is not helpful and can create loops of rooms/space switching
-                if (payload.context_switch) break;
+                if (!room || payload.context_switch) break;
 
                 // persist last viewed room from a space
 
-                // Don't save if the room is a space room. This would cause a problem:
-                // When switching to a space home, we first view that room and
-                // only after that we switch to that space. This causes us to
-                // save the space home to be the last viewed room in the home
-                // space.
-                if (room && !room.isSpaceRoom()) {
-                    window.localStorage.setItem(getLastViewedRoomsStorageKey(this.activeSpace), payload.room_id);
-                }
-
-                if (room?.getMyMembership() === "join") {
-                    if (room.isSpaceRoom()) {
-                        this.setActiveSpace(room);
-                    } else if (!this.spaceFilteredRooms.get(this._activeSpace?.roomId || HOME_SPACE).has(room.roomId)) {
-                        // TODO maybe reverse these first 2 clauses once space panel active is fixed
-                        let parent = this.rootSpaces.find(s => this.spaceFilteredRooms.get(s.roomId)?.has(room.roomId));
-                        if (!parent) {
-                            parent = this.getCanonicalParent(room.roomId);
-                        }
-                        if (!parent) {
-                            const parents = Array.from(this.parentMap.get(room.roomId) || []);
-                            parent = parents.find(p => this.matrixClient.getRoom(p));
-                        }
-                        if (parent) {
-                            this.setActiveSpace(parent);
-                        }
+                if (room.isSpaceRoom()) {
+                    this.setActiveSpace(room);
+                } else if (!this.getSpaceFilteredRoomIds(this.activeSpace).has(room.roomId)) {
+                    // TODO maybe reverse these first 2 clauses once space panel active is fixed
+                    let parent = this.rootSpaces.find(s => this.spaceFilteredRooms.get(s.roomId)?.has(room.roomId));
+                    if (!parent) {
+                        parent = this.getCanonicalParent(room.roomId);
+                    }
+                    if (!parent) {
+                        const parents = Array.from(this.parentMap.get(room.roomId) || []);
+                        parent = parents.find(p => this.matrixClient.getRoom(p));
+                    }
+                    if (parent) {
+                        // don't trigger a context switch when we are switching a space to match the chosen room
+                        this.setActiveSpace(parent, false);
```

Comment:
Shouldn't be there a fallback to the home space?

Reply:
That's a good point